### PR TITLE
[codex] add public contact payments

### PR DESCRIPTION
**Pull request by ben-kaufman**
### Description

Adds public Paykit payments to contacts on Android, matching the iOS public payments flow and the Paykit endpoint identifier spec.

- Publishes and removes spec-compliant Paykit payment endpoints for the current wallet.
- Adds payment discovery and launch flows for Add Contact and Contact Detail.
- Adds contact-tagged activity support and a contact activity screen.
- Threads contact context through immediate and pending send flows.
- Adds duplicate-contact validation parity and repo-level Paykit tests.
- Updates Paykit and bitkit-core bindings for the required endpoint and contact APIs.

### Preview

N/A

### QA Notes

Suggested reviewer checks:

- On the first-time Pay Contacts screen, choose enabled and continue; verify public endpoints are published for the current pubky.
- On a fresh first-time Pay Contacts path, choose disabled and continue; verify no public endpoints are published.
- Pay a saved contact with public endpoints from Add Contact and Contact Detail.
- Verify sent/pending contact payments attach activity to the contact.
- Verify Contact Activity shows `Sent to {name}` / `Received from {name}` titles.
- Paste or scan a pubky and verify routing to profile, existing contact, or add contact as appropriate.

### CODE REVIEW

**Comment by piotr-iohk** [comment]
Quick testing notes:
 - Was able to Send onchain to contact 👍 
 - Failed to send lightning to contact

[logs.zip](https://github.com/user-attachments/files/27375067/logs.zip)

https://github.com/user-attachments/assets/b10d0ce0-afe4-4e2e-8399-f40042b3e797



---

 - Send button should not be there on add contact screen, see: https://www.figma.com/design/ltqvnKiejWj0JQiqtDf2JJ/Bitkit-Wallet?node-id=40364-130299&t=xlSG6da1GwRIpMra-0
Screenshot 2026-05-04 at 22 05 32

---

- wrong icons and order for send and activity (iOS on the left, has correct ones) (https://www.figma.com/design/ltqvnKiejWj0JQiqtDf2JJ/Bitkit-Wallet?node-id=40364-130143&t=xlSG6da1GwRIpMra-0)
Screenshot 2026-05-04 at 22 06 32

---

- Contact activity only shows Sent to <Contact> - but does not show Received from <Contact>. Also on main activity "Received from..." is not displayed.(on both Android and iOS)
Screenshot 2026-05-04 at 22 10 52


---

 - The note on Add contact says: "Please note that you and Paolo must add each other as contacts to pay each other privately. Otherwise, the payment will be visible publicly." (https://www.figma.com/design/ltqvnKiejWj0JQiqtDf2JJ/Bitkit-Wallet?node-id=41985-298235&t=xlSG6da1GwRIpMra-0) - but you seem to be able to pay to contact even if he didn't add you as his contact.

**Comment by ovitrif** [comment]
@codex review

**Comment by ben-kaufman** [comment]
@piotr-iohk payment issue with LN seems like RGS error from the logs

File: app/src/main/java/to/bitkit/repositories/ActivityRepo.kt [hunk 1]
```diff
@@ -338,6 +342,74 @@ class ActivityRepo @Inject constructor(
...
+        runCatching {
+            val normalizedKey = PubkyPublicKeyFormat.normalized(publicKey) ?: publicKey
+            getActivities(
+                filter = ActivityFilter.ALL,
+                sortDirection = SortDirection.DESC,
+            ).getOrThrow().filter { activity ->
+                when (activity) {
+                    is Activity.Lightning -> PubkyPublicKeyFormat.matches(activity.v1.contact, normalizedKey)
+                    is Activity.Onchain -> PubkyPublicKeyFormat.matches(activity.v1.contact, normalizedKey)
+                }
```
Comment (ovitrif) [inline app/src/main/java/to/bitkit/repositories/ActivityRepo.kt]: nit: could have used the newly added `activity.contact()` ext.

File: app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    private val pubkyRepo: PubkyRepo,
+    private val walletRepo: WalletRepo,
+    private val lightningRepo: LightningRepo,
+    private val coreService: CoreService,
+    private val settingsStore: SettingsStore,
+    private val clock: Clock,
+) {
+    companion object {
+        private val methodIdPattern = Regex("^[a-z0-9]+-[a-z0-9]+-[a-z0-9]+$")
+        private val json = Json { ignoreUnknownKeys = true }
```
Comment (ovitrif) [inline app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt]: nit: repeating this instantiation of Kotlin's `Json` every time is going to slow down the app in the long run, usual recommendation is to rely on as few instances of it as possible. I know this because I still open the repo using Android Studio, and it usually complains where it sees this.

we have `json` global object in `JsonModule`. Maybe it's re-usable here, out-of-the-box? If not, maybe AI knows a better approach to push for more reuse of JSON instances.

PS. Could also add a rule to AGENTS.md so this is known going forwards, that we should prefer to reuse existing `Json` instance objects.

File: app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            MethodId.Bolt11,
+            MethodId.Lnurl,
+            MethodId.P2tr,
+            MethodId.P2wpkh,
+            MethodId.P2sh,
+            MethodId.P2pkh,
+        )
+
+        private val managedMethodIds = MethodId.entries.filter { it.isBitkitManaged }
+        private val publicBolt11Expiry = 24.hours
```
Comment (ovitrif) [inline app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt:79]: nit: should we sync this with the default lifetime of bolt11 invoices created via LDK? For a better experience & optimal coding, (idk?!, need 2 brains on this)

see
```kt
suspend fun receive(sat: ULong? = null, description: String, expirySecs: UInt = 3600u)

// or
suspend fun receiveMsats(amountMsat: ULong? = null, description: String, expirySecs: UInt = 3600u)
```

File: app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        return currentPublicKey
+    }
+
+    private suspend fun buildWalletEndpoints(refresh: Boolean): List<Endpoint> {
+        if (refresh) {
+            lightningRepo.executeWhenNodeRunning(
+                operationName = "sync public Paykit endpoints",
+            ) {
+                Result.success(Unit)
+            }.getOrThrow()
```
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt:244]: Then maybe we can wrap it at callsite in `runCatching`. wdyt @codex?

File: app/src/main/java/to/bitkit/data/SettingsStore.kt [hunk 5]
```diff
@@ -100,6 +100,11 @@ data class SettingsData(
     val hasSeenShopIntro: Boolean = false,
     val hasSeenProfileIntro: Boolean = false,
     val hasSeenContactsIntro: Boolean = false,
+    val hasConfirmedPublicPaykitEndpoints: Boolean = false,
```
Comment (ovitrif) [inline app/src/main/java/to/bitkit/data/SettingsStore.kt:103]: Since this adds new persisted settings fields, do we need to update the settings backup/restore model as well, or is this already covered because `SettingsBackupV1` stores the full `SettingsData`?

File: app/src/main/java/to/bitkit/ui/screens/wallets/activity/components/ContactActivityTitle.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import com.synonym.bitkitcore.Activity
+import kotlinx.collections.immutable.ImmutableList
+import to.bitkit.R
+import to.bitkit.ext.contact
+import to.bitkit.ext.isSent
+import to.bitkit.models.PubkyProfile
+import to.bitkit.models.PubkyPublicKeyFormat
+
+@Composable
+fun contactActivityTitle(activity: Activity, contacts: ImmutableList<PubkyProfile>): String? {
```
Comment (ovitrif) [inline app/src/main/java/to/bitkit/ui/screens/wallets/activity/components/ContactActivityTitle.kt:15]: nit: file name should usually reflect name of single code construct if there's only one in the file; or in applicable cases, the main construct described within it.

In this case, best to name file same like the method: `ContactActivityTitle`.

Rule is usually more relevant for classes, etc; but also transmutable to fns.

File: app/src/test/java/to/bitkit/repositories/PublicPaykitRepoTest.kt [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+    companion object {
+        private const val NOW_MILLIS = 1_000L
+    }
+
+    @Test
+    fun `syncCurrentPublishedEndpoints sets desired endpoints and removes obsolete bitkit endpoints`() = test {
+        val pubkyRepo = mock<PubkyRepo>()
+        val walletRepo = mock<WalletRepo>()
+        val lightningRepo = mock<LightningRepo>()
+        val coreService = mock<CoreService>()
```
Comment (ovitrif) [inline app/src/test/java/to/bitkit/repositories/PublicPaykitRepoTest.kt]: nit: maybe ask it to extract props for test class, so we don't have to reinit every time, makes test bodies shorter; if that works out-of-the-box as a change in this test class.

**Review by ovitrif — COMMENTED** [review]
LGTM code-wise

Very few nits considering the amount of changes (+2k lines), congrats 👏🏻 .

Going to test some edge cases, mainly those which reflect production use.

Thinking of:
- [x] existing wallet without paykit, getting paykit then participating in pay2contact public
- [x] wallet without LN getting paykit profile, acquiring LN channel, then trying to receive/send
- [ ] mainnet pay2contact public <> (to ensure demo ready- for Denys)

overall this ↑ kind of permutations. cc @piotr-iohk depending on how we want to merge & test after merge, if we want everything to be ready before merge, which I fully support, then we need to be sure these cases, and more along this lines that we can think of, work great.

File: app/src/main/java/to/bitkit/data/SettingsStore.kt (hunk 5 above)
Comment (ovitrif) [reply app/src/main/java/to/bitkit/data/SettingsStore.kt:103]: We don't, I asked Codex.

File: app/src/main/java/to/bitkit/viewmodels/AppViewModel.kt [hunk 8]
```diff
@@ -1296,6 +1364,21 @@ class AppViewModel @Inject constructor(
...
+            activeContactPaymentContext = ContactPaymentContext(publicKey)
+        }
+        onScanResult(paymentRequest)
+    }
+
+    fun preserveContactPaymentContext(paymentHash: String) {
+        synchronized(contactPaymentContextLock) {
+            activeContactPaymentContext?.let {
+                pendingContactPaymentContexts[paymentHash] = it
+            }
```
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/viewmodels/AppViewModel.kt]: @codex how to "Move the context instead of copying it" here? can you suggest code change directly in GitHub?

**Comment by piotr-iohk** [comment]
> @piotr-iohk payment issue with LN seems like RGS error from the logs

@ben-kaufman, pls note that I'm able to send bolt11 to the recepient address, but the issue is only when sending to contact. (This is on latest revision afeb1f47e8fa3cb48e25ae3137d0f7f784fffba6 and staging env). I don't see this problem on iOS.

[logs.zip](https://github.com/user-attachments/files/27392819/logs.zip)

https://github.com/user-attachments/assets/d34019a5-26c9-4ddf-b090-d748fe8103bd

**Comment by piotr-iohk** [comment]
> mainnet pay2contact public <> (to ensure demo ready- for Denys)

Just to note, profile creation will not work with mainnet build until this probably https://github.com/pubky/pubky-stack/pull/243

File: app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt:244]: @codex, can you answer the questio. now, is it ok to just wrap the call to this method inside `runCatching {./* here */ }`? this is a private method, not public.

File: app/src/main/java/to/bitkit/viewmodels/AppViewModel.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/viewmodels/AppViewModel.kt]: @codex please try I again, I set that up.

File: app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/repositories/PublicPaykitRepo.kt:244]: Dude, node must run anyways. It's gateway to onchain payments too.

**Comment by piotr-iohk** [comment]
Tested roughly against this [plan](https://github.com/synonymdev/bitkit-e2e-tests/blob/feat/pubky-profile/docs/public-contact-payments-manual-qa.md). After some time my issue with ln payment disappeared - so in the end I was able to send lightning to contact - still may be worth looking into it. Have some more testing notes below, but mostly that's polishing. Tested against staging.

Areas tested and 🟢 for me.
 - send/receive with imported pubky to saved/unsaved contact (onchain and lightning)
 - send/receive with created pubky to saved/unsaved contact (onchain and lightning)
 - RBF sent tx to concat
 - CPFP on received tx from contact
 - enable/disable payments with contacts

Some testing notes / observations:

 - adding/paying contact via paste invoice glitch - add contact screen "in bg":

https://github.com/user-attachments/assets/9aec5c04-b06e-4f3f-9817-721d488a5843

--- 

 - Enter manually is not supporting pubky (since we have it supported on paste/qr code scan etc. - then Send -> Enter manually should also support it)
Screenshot 2026-05-05 at 13 27 20

---

- RBF-ed transaction not shown in profile activity (Instead there is a sent tx with Removed from Mempol note)
   - to reproduce:
   - send to contact and RBF transaction
   - see transactions in contact activity
Screenshot 2026-05-05 at 13 30 01

**Review by ovitrif — DISMISSED** [review]
LGTM

Didn't encounter any other issue except for the ones that Piotr mentioned, and the pay to iOS not working in that exceptional case when actually the issue was on the iOS side. Once I 'restored' my profile on iOS, paying to the contact via Android worked.

---

PS. Recording of how send 2 contact via Send → Paste behaves (issue reported by Piotr):
https://github.com/user-attachments/assets/2fa3a110-a80c-47b2-861a-65f173252684

**Review by piotr-iohk — DISMISSED** [review]
LGTM, remaining obesrvations can be addressed separately, issue for tracking: https://github.com/synonymdev/bitkit-android/issues/931

**Review by ovitrif — DISMISSED** [review]
re-approved after child PR got merged:
- #930

**Comment by ovitrif** [comment]
Created issue for tracking fixing of the [remarks from this review comment](https://github.com/synonymdev/bitkit-android/pull/924#issuecomment-4379223065):
- https://github.com/synonymdev/bitkit-android/issues/932

File: app/src/main/java/to/bitkit/services/PubkyService.kt [hunk 9]
```diff
@@ -18,26 +19,38 @@ import com.synonym.bitkitcore.pubkySessionPut
...
-class PubkyService @Inject constructor() {
+class PubkyService @Inject constructor(
+    @ApplicationContext private val context: Context,
+) {
 
     private val isSetup = CompletableDeferred<Unit>()
 
     suspend fun initialize() = ServiceQueue.CORE.background {
+        if (!PaykitAndroid.initialize(context)) {
+            throw AppError("Failed to initialize Android platform verifier")
```
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/services/PubkyService.kt:52]: Nah, who cares?!

File: app/src/main/java/to/bitkit/repositories/LightningRepo.kt [hunk 10]
```diff
@@ -917,7 +918,7 @@ class LightningRepo @Inject constructor(
     suspend fun createInvoice(
         amountSats: ULong? = null,
         description: String,
-        expirySeconds: UInt = 86_400u,
+        expirySeconds: UInt = Defaults.bolt11InvoiceExpirySeconds,
```
[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/repositories/LightningRepo.kt]: Point seems relevant, it slipped through unintentionally in child PR.
I only meant to change it further down the pipeline.

Now I wonder, @claude, this is actually how invoices get generated, always with 24h when user is using `Receive` flow?

Isn't the code quite confusing, because the default in LightningService uses default of 1h…, while main use-case for users goes through this method, thus most of the time, effective default duration is 24h?

Please check what other cases still end up actually using the default of 1h.

Seems we need to update the default to 24h.

[the remark this answers is not in the document]
Comment (ovitrif) [reply app/src/main/java/to/bitkit/repositories/LightningRepo.kt]: NVM Claude usage limits nowadays, I verified by testing vs. iOS, this should be reverted.

**Review by ovitrif — APPROVED** [review]
re-reviewed.

Thx for the fixes 🙌🏻

**Comment by ovitrif** [comment]
Merging this to master 🎉 🚀 .

Thanks team 🙋🏻
